odxtools/parameters/tablekeyparameter.py

```diff
@@ -37,37 +40,107 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
         super()._resolve_odxlinks(odxlinks)
 
         if self.table_ref:
-            self._table = odxlinks.resolve(self.table_ref)
+            if TYPE_CHECKING:
+                self._table = odxlinks.resolve(self.table_ref, Table)
+            else:
+                self._table = odxlinks.resolve(self.table_ref)
+
+        self._table_row: Optional[TableRow] = None
         if self.table_row_ref:
-            self._table_row = odxlinks.resolve(self.table_row_ref)
+            if TYPE_CHECKING:
+                self._table_row = odxlinks.resolve(self.table_row_ref, TableRow)
+            else:
+                self._table_row = odxlinks.resolve(self.table_row_ref)
 
     def _resolve_snrefs(self, diag_layer: "DiagLayer") -> None:
         super()._resolve_snrefs(diag_layer)
 
         if self.table_snref:
-            self._table = diag_layer.tables[self.table_snref]
+            ddd_spec = diag_layer.diag_data_dictionary_spec
+            self._table = ddd_spec.tables[self.table_snref]
         if self.table_row_snref:
-            self._table_row = diag_layer.tables[self.table_row_snref]
+            # make sure that we know the table to which the table row
+            # SNREF is relative to.
+            assert self._table is not None, "If a table-row short name " \
+                "reference is defined, a table must also be specified."
+            self._table_row = self._table.table_rows[self.table_row_snref]
 
     @property
     def table(self) -> "Table":
         return self._table
 
     @property
-    def table_row(self) -> "TableRow":
+    def table_row(self) -> Optional["TableRow"]:
         return self._table_row
 
     def is_required(self):
-        raise NotImplementedError("TableKeyParameter.is_required is not implemented yet.")
+        return self._table_row is None
```

Comment:
It should be possible to figure out the value for this from the TableStructParameter, same way we figure out the length parameter, no?

Reply:
not sure what you mean: IIRC, `is_required()` says whether the parameter must be specified for `encode()` or if it can be defaulted or not. this is the the case iff we are not in the static case, i.e., the table row is not defined.

Reply:
It is still possible to figure out the tablekeyparameter value from the tablestructparameter, same way we figure out the lengthkeyparameter, no?

Reply:
not really: for this, we would need to know what struct was selected and to determine this we would need the assumption that the key->struct mapping is one-to-one, i.e. that given a struct, we can find a unique key (this is not the case since tables are free to map multiple keys to the same struct).

The reason why we can figure out the value of length keys implicitly while encoding is that they are just the length of the list of the associated parameter...

Reply:
Each row have a short_name, the idea is that instead of the tablestructparameter containing directly the struct, it would be wrapped in a dict with one key-value, key being the row short_name, and value being the struct that needs to be encoded

This would be simular to how we encode mutex at the moment.

Reply:
do you mean to encode PDUs with table-struct parameters like this:
```python
rq.encode(table_struct_param={"nice_table_row_key":{"table_struct_param1":"table_struct_value1", }})
```

instead of
```python
rq.encode(table_key_param="nice_table_row_key", table_struct_param={"table_struct_param1":"table_struct_value1", })
```
?

the issue with this is that ODX separate the encoding of the table-keys from the table structs, and there is no "back link" from the table keys to the table structs. This means that (in principle), multiple table struct parameters can use the same table key. (I doubt that many people use this, but ...). Of course, these issues can be solved, but I wonder if they are worth it. Possibly, it is better to remove the implicit encoding of length-keys instead...

Reply:
> do you mean to encode PDUs with table-struct parameters like this

Yes

Let me explain the reasons for my request a bit
* Having to specify two parameters independently will lead to weird error messages when user change one and forget to change the other
* I am (in my internal projects) trying to generate JSON schema from the PDX parameters (and use it to generate a GUI), so having a consitent behavior for each parameter (and not chaning what it takes as value) would be nicer

> there is no "back link" from the table keys to the table structs. This means that (in principle), multiple table struct parameters can use the same table key.

That's another point, people who have those "weird" uses cases, can do this
```
rq.encode(table_key_param=table_row_key_physical_value, table_struct_param={"table_row_short_name":{"table_struct_param1":"table_struct_value1", }})
```
table_key_param will have a default value from table_struct_param (the same way we encode length paramters)
but user can still specify it explicitly